### PR TITLE
Create MY_OBZ_Demo_Amirreza_Silani

### PR DESCRIPTION
**Pull request by asilani**

### CODE REVIEW

File: MY_OBZ_Demo_Amirreza_Silani [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (asilani) [inline MY_OBZ_Demo_Amirreza_Silani:1]: [MY_OBZ_Demo_Amirreza_Silani.zip](https://github.com/user-attachments/files/20704063/MY_OBZ_Demo_Amirreza_Silani.zip)

Comment (asilani) [inline MY_OBZ_Demo_Amirreza_Silani:1]: #Guarantee to run thew current directory
# using Pkg: Pkg       # Julia package manager (like pip for Python)
# Pkg.activate(".")    # Creates and activates the project in the new folder - notice it creates Project.toml and Manifest.toml in your folder for reproducibility

# # Load the packages
# import TulipaIO as TIO
# import TulipaEnergyModel as TEM
# using DuckDB
# using DataFrames
# using Plots

# connection = DBInterface.connect(DuckDB.DB)
# input_dir = "my-energy-system1"
# output_dir = "my-energy-system-results1"

# TIO.read_csv_folder(connection, input_dir)

# # Explore the tables in DuckDB
# TIO.show_tables(connection)
# TIO.get_table(connection, "asset_milestone")


# TEM.populate_with_defaults!(connection)

# # Explore the tables in DuckDB (again)
# TIO.show_tables(connection)
# TIO.get_table(connection, "asset_milestone")

# energy_problem =
#     TEM.run_scenario(connection; output_folder=output_dir)

# guaranteed to be run in the current directory
# using Pkg: Pkg
# Pkg.activate(".")

# # load the packages
# import TulipaIO as TIO
# import TulipaEnergyModel as TEM
# using DuckDB
# using DataFrames
# using Plots

# Define the directories
# input_dir = "my-energy-system2"
# output_dir = "my-energy-system-results2"

# # Create the connection and read the case study files
# connection = DBInterface.connect(DuckDB.DB)
# TIO.read_csv_folder(connection, input_dir)

# # Add the defaults
# TEM.populate_with_defaults!(connection)

# # Optimize the model
# energy_problem =
#     TEM.run_scenario(connection; output_folder=output_dir)


# guaranteed to be run in the current directory
# guaranteed to be run in the current directory
# using Pkg: Pkg
# Pkg.activate(".")

# # load the packages
# import TulipaIO as TIO
# import TulipaEnergyModel as TEM
# using DuckDB
# using DataFrames
# using Plots

# # Define the directories
# input_dir = "my-energy-system3"
# output_dir = "my-energy-system-results2"

# # Temporary fix!!: pass the schema of the partition files
# schema_partition_files = Dict(
#     table_name => TEM.schema_per_table_name[table_name]
#     for table_name in ["assets_rep_periods_partitions", "flows_rep_periods_partitions"]
# )

# # Create the connection and read the case study files
# connection = DBInterface.connect(DuckDB.DB)
# TIO.read_csv_folder(connection, input_dir; schemas=schema_partition_files)

# # Add the defaults
# TEM.populate_with_defaults!(connection)

# # Optimize the model
# energy_problem =
#     TEM.run_scenario(connection; output_folder=output_dir)




# using DataFrames
# using Plots

# flows = TIO.get_table(connection, "var_flow")

# from_asset = "wind"
# to_asset = "e_demand"
# year = 2030
# rep_period = 1


# filtered_flow = filter(
#     row ->
#         row.from_asset == from_asset &&
#             row.to_asset == to_asset &&
#             row.year == year &&
#             row.rep_period == rep_period,
#     flows,
# )

# plot(
#     filtered_flow.time_block_start,
#     filtered_flow.solution;
#     label=string(from_asset, " -> ", to_asset),
#     xlabel="Hour",
#     ylabel="[MWh]",
#     dpi=600,
# )


# balance = TIO.get_table(connection, "cons_balance_consumer")

# asset = "e_demand"
# year = 2030
# rep_period = 1

# filtered_asset = filter(
#     row ->
#         row.asset == asset &&
#             row.year == year &&
#             row.rep_period == rep_period,
#     balance,
# )


# plot(
#     filtered_asset.time_block_start,
#     filtered_asset.dual_balance_consumer;
#     #label=string(from_asset, " -> ", to_asset),
#     xlabel="Hour",
#     ylabel="[MWh]",
#     #xlims=(0, 168),
#     dpi=600,
# )



# transport = TIO.get_table(connection, "cons_transport_flow_limit_simple_method")

# names(transport)

# filter(
#     row ->
#         row.dual_max_transport_flow_limit_simple_method != 0.0,
#     transport,
# )





# flows = TIO.get_table(connection, "var_flow")

# from_asset = "electrolizer"
# to_asset = "h2_demand"
# year = 2030
# rep_period = 1

# filtered_flow = filter(
#     row ->
#         row.from_asset == from_asset &&
#             row.to_asset == to_asset &&
#             row.year == year &&
#             row.rep_period == rep_period,
#     flows,
# )

# plot(
#     filtered_flow.time_block_start,
#     filtered_flow.solution;
#     label=string(from_asset, " -> ", to_asset),
#     xlabel="Hour",
#     ylabel="[MWh]",
#     marker=:circle,
#     markersize=2,
#     linetype=:steppre, # try: stepmid, steppost, or steppre
#     xlims=(1, 168 * 4),
#     dpi=600,
# )



# balance = TIO.get_table(connection, "cons_balance_consumer")

# asset = "h2_demand"
# year = 2030
# rep_period = 1

# filtered_asset = filter(
#     row ->
#         row.asset == asset &&
#             row.year == year &&
#             row.rep_period == rep_period,
#     balance,
# )



# flows = TIO.get_table(connection, "var_flow")

# from_asset = "electrolizer"
# to_asset = "h2_demand"
# year = 2030
# rep_period = 1

# filtered_flow = filter(
#     row ->
#         row.from_asset == from_asset &&
#             row.to_asset == to_asset &&
#             row.year == year &&
#             row.rep_period == rep_period,
#     flows,
# )

# plot(
#     filtered_flow.time_block_start,
#     filtered_flow.solution;
#     label=string(from_asset, " -> ", to_asset),
#     xlabel="Hour",
#     ylabel="[MWh]",
#     marker=:circle,
#     markersize=2,
#     linetype=:steppre, # try: stepmid, steppost, or steppre
#     xlims=(1, 168 * 4),
#     dpi=600,
# )

# balance = TIO.get_table(connection, "cons_balance_consumer")

# asset = "h2_demand"
# year = 2030
# rep_period = 1

# filtered_asset = filter(
#     row ->
#         row.asset == asset &&
#             row.year == year &&
#             row.rep_period == rep_period,
#     balance,
# )


# conn_hourly = DBInterface.connect(DuckDB.DB)
# input_dir = "my-energy-system2"
# TIO.read_csv_folder(conn_hourly, input_dir)
# TEM.populate_with_defaults!(conn_hourly)
# hourly_energy_problem = TEM.run_scenario(conn_hourly)



# flows = TIO.get_table(connection, "var_flow")
# from_asset = "electrolizer"
# to_asset = "h2_demand"
# year = 2030
# rep_period = 1

# filtered_flow = filter(
#     row ->
#         row.from_asset == from_asset &&
#             row.to_asset == to_asset &&
#             row.year == year &&
#             row.rep_period == rep_period,
#     flows,
# )

# plot(
#     filtered_flow.time_block_start,
#     filtered_flow.solution;
#     label=string(from_asset, " -> ", to_asset),
#     xlabel="Hour",
#     ylabel="[MWh]",
#     marker=:circle,
#     markersize=2,
#     linetype=:steppre, # try: stepmid, steppost, or steppre
#     xlims=(1, 168 * 4),
#     dpi=600,
# )

# hourly_flows = TIO.get_table(conn_hourly, "var_flow")

# hourly_filtered_flow = filter(
#     row ->
#         row.from_asset == from_asset &&
#             row.to_asset == to_asset &&
#             row.year == year &&
#             row.rep_period == rep_period,
#     hourly_flows,
# )

# plot!(
#     hourly_filtered_flow.time_block_start,
#     hourly_filtered_flow.solution;
#     label=string(from_asset, " -> ", to_asset, " (hourly)"),
#     xlabel="Hour",
#     ylabel="[MWh]",
#     xlims=(1, 168 * 4),
#     dpi=600,
# )



########## Session 4:

using Pkg
Pkg.activate(".")
# Pkg.add("TulipaEnergyModel")
# Pkg.add("TulipaIO")
Pkg.add("TulipaClustering") # NB: this is new; ask students if they remember how to install a package
# Pkg.add("DuckDB")
# Pkg.add("DataFrames")
# Pkg.add("Plots")
Pkg.add("Distances")

Pkg.instantiate()

import TulipaIO as TIO
import TulipaEnergyModel as TEM
import TulipaClustering as TC  # NB: this is new
using DuckDB
using DataFrames
using Plots
using Distances  # NB: this is new

connection = DBInterface.connect(DuckDB.DB)

input_dir = "my-energy-system"
output_dir = "my-energy-system-results"

TIO.read_csv_folder(connection, input_dir)


period_duration = 24
profiles_df = TIO.get_table(connection, "profiles_periods")
TC.combine_periods!(profiles_df)
TC.split_into_periods!(profiles_df; period_duration)

num_rep_periods = 2
method = :k_medoids  # :k_means, :convex_hull, :convex_hull_with_null, :conical_hull
distance = Euclidean()  # CosineDist()

clustering_result = TC.find_representative_periods(profiles_df, num_rep_periods; method, distance)

weight_type = :dirac  # :convex, :conical, :conical_bounded
tol = 1e-2
niters = 100
learning_rate = 0.001

TC.fit_rep_period_weights!(
    clustering_result;
    weight_type,
    tol,
    niters,
    learning_rate,
)
TC.write_clustering_result_to_tables(connection, clustering_result)

TEM.populate_with_defaults!(connection)
energy_problem = TEM.run_scenario(connection; output_folder=output_dir)


flows = TIO.get_table(connection, "var_flow")
select!(
    flows,
    :from_asset,
    :to_asset,
    :year,
    :rep_period,
    :time_block_start => :timestep,
    :solution
)

from_asset = "ccgt"
to_asset = "e_demand"
year = 2030

filtered_flow = filter(
    row ->
        row.from_asset == from_asset &&
            row.to_asset == to_asset &&
            row.year == year,
    flows,
)


rep_periods_mapping = TIO.get_table(connection, "rep_periods_mapping")

df = innerjoin(filtered_flow, rep_periods_mapping, on=[:year, :rep_period])
gdf = groupby(df, [:from_asset, :to_asset, :year, :period, :timestep])
result_df = combine(gdf, [:weight, :solution] => ((w, s) -> sum(w .* s)) => :solution)
TC.combine_periods!(result_df)
sort!(result_df, :timestep)
plot(
    result_df.timestep,
    result_df.solution;
    label=string(from_asset, " -> ", to_asset),
    xlabel="Hour",
    ylabel="[MWh]",
    marker=:circle,
    markersize=2,
    xlims=(1, 168),
    dpi=600,
)
